flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/classification/MultinomialNaiveBayes.scala

```diff
@@ -0,0 +1,900 @@
+          } // (class name -> word -> count of that word -> count of all words in that class)
+
+          //calculate the P(w|c) value for each word in each class
+          // 1. Map: use normal P(w|c) formula
+          pwc = wordsInClass.map(line => (line._1, line._2, (line._3 + 1) /
+            (line._4 + vocabularyCount)))
+
+      } else if (p3 == 1) {
+
+        //calculate the P(w|c) value for each word in class
+        //  1. Map: use normal P(w|c) formula / use the
+        pwc = singleWordsInClass.map(new RichMapFunction[(String, String, Int),
+          (String, String, Double)] {
+
+          var broadcastMap: mutable.Map[String, Int] = mutable.Map[String, Int]()
+
+
+          override def open(config: Configuration): Unit = {
+            val collection = getRuntimeContext
+              .getBroadcastVariable[(String, Int)]("allWordsInClass").asScala
+            for (record <- collection) {
+              broadcastMap.put(record._1, record._2)
+            }
+          }
+
+          override def map(value: (String, String, Int)): (String, String, Double) = {
+            (value._1, value._2, (value._3 + 1) / (broadcastMap(value._1) + vocabularyCount))
+          }
+        }).withBroadcastSet(allWordsInClass, "allWordsInClass")
+
+      }
+
+      //END POSSIBILITY 3
+
+      //stores all the word related information in one data set
+      // 1. Map: Caluclate logarithms
+      val wordRelatedModelData = pwc.map(line => (line._1, line._2, Math.log(line._3)))
+
+      //store all class related information in one data set
+      // 1. Join: P(c) data set and P(w|c) data set not in class and calculate logarithms
+      val classRelatedModelData = pc.join(pwcNotInClass)
+        .where(0).equalTo(0) {
+        (line1, line2) => (line1._1, Math.log(line1._2), Math.log(line2._2))
+      } // (class name -> log(P(c)) -> log(P(w|c) not in class))
+
+      instance.wordRelatedModelData = Some(wordRelatedModelData)
+      instance.classRelatedModelData = Some(classRelatedModelData)
+
+      //RENNIE 1: transform document frequency
+        //for this, the improvementData set must be set
+        //calculate (word -> log number of documents in all classes / docs with that word)
+
+      val r1 = resultingParameters(R1)
+
+      if (r1 == 1) {
+        val totalDocumentCount: DataSet[(Double)] = documentsPerClass
+          .reduce((line1, line2) => (line1._1, line1._2 + line2._2))
+          .map(line => line._2) // (count of all documents)
+
+        //number of occurences over all documents of all classes
+        val wordCountTotal = input
+          .flatMap(new SingleDistinctWordSplitter())
+          .map(line => (line._2, 1))
+          .groupBy(0)
+          .reduce((line1, line2) => (line1._1, line1._2 + line2._2))
+           // (word -> count of documents with that word)
+
+        val improvementData = wordCountTotal.map(new RichMapFunction[(String, Int),
```

Comment:
You're broadcasting a set. At some point, the following code will be there in `ml/package.scala`
```scala
def mapWithBcSet[B, O: TypeInformation: ClassTag](
    broadcastVariable: DataSet[B])(
    fun: (T, Seq[B]) => O)
  : DataSet[O] = {
  dataSet.map(new BroadcastSetMapper[T, B, O](dataSet.clean(fun)))
  .withBroadcastSet(broadcastVariable, "broadcastVariable")
}

private class BroadcastSetMapper[T, B, O](fun: (T, Seq[B]) => O)
  extends RichMapFunction[T, O] {

  var broadcastVariable: Seq[B] = _

  @throws(classOf[Exception])
  override def open(configuration: Configuration): Unit = {
    broadcastVariable = getRuntimeContext
      .getBroadcastVariable[B]("broadcastVariable")
      .asScala
      .toSeq
  }

  override def map(value: T): O = {
    fun(value, broadcastVariable)
  }
}
```
You can include this in your code or copy this file: https://raw.githubusercontent.com/sachingoel0101/flink/4e4e10259a48f3bbcbcd48410fd01d45d4882f2c/flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/package.scala

This will provide a much neater way to broadcast sets.